File: iris-ld/iris-ld-proofs/src/main/java/city/smartb/iris/ldproof/verifier/LdProofVerifier.java
package city.smartb.iris.ldproof.verifier;

import city.smartb.iris.crypto.dsl.verifier.Verifier;
import city.smartb.iris.ldproof.VerifiableJsonLd;
import city.smartb.iris.ldproof.VerifiableJsonLdBuilder;
import city.smartb.iris.ldproof.LdProof;
import city.smartb.iris.ldproof.LdProofBuilder;
import city.smartb.iris.ldproof.util.JWSUtil;
import city.smartb.iris.ldproof.util.SHAUtil;
import com.nimbusds.jose.JOSEException;
import com.nimbusds.jose.JWSAlgorithm;
import com.nimbusds.jose.JWSObject;
import com.nimbusds.jose.JWSVerifier;

import java.security.GeneralSecurityException;
import java.text.ParseException;
import java.util.Map;

public abstract class LdProofVerifier {

    private Verifier verifier;

    protected LdProofVerifier(Verifier verifier) {
        this.verifier = verifier;
    }

    public boolean verify(VerifiableJsonLd jsonLdWithProof) throws GeneralSecurityException {
        return verify(jsonLdWithProof.asJson());
    }

    public boolean verify(Map<String, Object> jsonLdObject) throws GeneralSecurityException {
        VerifiableJsonLd ldObject = new VerifiableJsonLd(jsonLdObject);
        LdProof proofs = ldObject.getProof();
        return this.verify(jsonLdObject, proofs);
    }

   public boolean verify(Map<String, Object> jsonLdObject, LdProof ldProof) throws GeneralSecurityException {

        LdProofBuilder builder = LdProofBuilder.fromLdProof(ldProof);
        String canonicalizedProof = builder.canonicalize();

        String canonicalizedDocument = VerifiableJsonLdBuilder.builder(jsonLdObject).buildCanonicalizedDocument();

        byte[] signingInput = new byte[64];
        System.arraycopy(SHAUtil.sha256(canonicalizedProof), 0, signingInput, 0, 32);
        System.arraycopy(SHAUtil.sha256(canonicalizedDocument), 0, signingInput, 32, 32);

        boolean verify = this.verify(signingInput, ldProof);

        // done

        return verify;
    }

    private boolean verify(byte[] signingInput, LdProof ldProof) throws GeneralSecurityException {
        try {
            String jws = ldProof.getJws();
            JWSObject detachedJwsObject = JWSObject.parse(jws);
            byte[] jwsSigningInput = JWSUtil.getJwsSigningInput(detachedJwsObject.getHeader(), signingInput);
            JWSVerifier jwsVerifier = new JoseJWSVerifier(verifier, JWSAlgorithm.RS256);
            return jwsVerifier.verify(detachedJwsObject.getHeader(), jwsSigningInput, detachedJwsObject.getSignature());

        } catch (JOSEException | ParseException ex) {
            throw new GeneralSecurityException("JOSE verification problem: " + ex.getMessage(), ex);
        }
    }

}

File: iris-ld/iris-ld-proofs/src/main/kotlin/city/smartb/iris/ld/ldproof/verifier/LdProofVerifier.kt
package city.smartb.iris.ld.ldproof.verifier

import city.smartb.iris.crypto.dsl.verifier.Verifier
import city.smartb.iris.ld.ldproof.LdProof
import city.smartb.iris.ld.ldproof.LdProofBuilder
import city.smartb.iris.ld.ldproof.VerifiableJsonLd
import city.smartb.iris.ld.ldproof.VerifiableJsonLdBuilder
import city.smartb.iris.ld.ldproof.util.JWSUtil
import city.smartb.iris.ld.ldproof.util.SHAUtil
import com.nimbusds.jose.JOSEException
import com.nimbusds.jose.JWSAlgorithm
import com.nimbusds.jose.JWSObject
import com.nimbusds.jose.JWSVerifier
import java.security.GeneralSecurityException
import java.text.ParseException

abstract class LdProofVerifier protected constructor(private val verifier: Verifier) {
    @Throws(GeneralSecurityException::class)
    fun verify(jsonLdWithProof: VerifiableJsonLd): Boolean {
        return verify(jsonLdWithProof.asJson())
    }

    @Throws(GeneralSecurityException::class)
    fun verify(jsonLdObject: Map<String, Any>): Boolean {
        val ldObject = VerifiableJsonLd(jsonLdObject)
        val proofs = ldObject.proof
        return this.verify(jsonLdObject, proofs)
    }

    @Throws(GeneralSecurityException::class)
    fun verify(
        jsonLdObject: Map<String, Any>,
        ldProof: LdProof
    ): Boolean {
        val jws = ldProof.jws!!
        val builder: LdProofBuilder = LdProofBuilder.Companion.fromLdProof(ldProof)
        val canonicalizedProof = builder.canonicalize()
        val canonicalizedDocument: String =
            VerifiableJsonLdBuilder.Companion.builder(jsonLdObject).buildCanonicalizedDocument()
        val signingInput = ByteArray(64)
        System.arraycopy(SHAUtil.sha256(canonicalizedProof), 0, signingInput, 0, 32)
        System.arraycopy(SHAUtil.sha256(canonicalizedDocument), 0, signingInput, 32, 32)

        // done
        return verify(signingInput, jws)
    }

    @Throws(GeneralSecurityException::class)
    private fun verify(signingInput: ByteArray, jws: String): Boolean {
        return try {
            val detachedJwsObject = JWSObject.parse(jws)
            val jwsSigningInput = JWSUtil.getJwsSigningInput(detachedJwsObject.header, signingInput)
            val jwsVerifier: JWSVerifier = JoseJWSVerifier(verifier, JWSAlgorithm.RS256)
            jwsVerifier.verify(detachedJwsObject.header, jwsSigningInput, detachedJwsObject.signature)
        } catch (ex: JOSEException) {
            throw GeneralSecurityException("JOSE verification problem: " + ex.message, ex)
        } catch (ex: ParseException) {
            throw GeneralSecurityException("JOSE verification problem: " + ex.message, ex)
        }
    }
}
